test/general/implementation_details/indirectly_device_accessible.pass.cpp

```diff
@@ -0,0 +1,464 @@
+
+    // test custom user second strided iterator (no body for is_onedpl_indirectly_device_accessible_iterator)
+    using SecondStridedIter = custom_user::second_strided_iterator<BaseIter>;
+    static_assert(oneapi::dpl::is_indirectly_device_accessible_v<SecondStridedIter> ==
+                      base_indirectly_device_accessible,
+                  "is_onedpl_indirectly_device_accessible_iterator is not working correctly for custom user strided "
+                  "iterator with no body in ADL function definition");
+
+    // test custom user first strided iterator with hidden friend ADL function
+    using ThirdStridedIter = custom_user::third_strided_iterator<BaseIter>;
+    static_assert(oneapi::dpl::is_indirectly_device_accessible_v<ThirdStridedIter> == base_indirectly_device_accessible,
+                  "is_onedpl_indirectly_device_accessible_iterator is not working correctly for custom user strided "
+                  "iterator with hidden friend ADL function");
+
+    // test custom user first strided iterator with hidden friend ADL function without body
+    using FourthStridedIter = custom_user::fourth_strided_iterator<BaseIter>;
+    static_assert(oneapi::dpl::is_indirectly_device_accessible_v<FourthStridedIter> ==
+                      base_indirectly_device_accessible,
+                  "is_onedpl_indirectly_device_accessible_iterator is not working correctly for custom user strided "
+                  "iterator with hidden friend ADL function without a body");
+}
+
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+int
+main()
+{
+
+#if TEST_DPCPP_BACKEND_PRESENT
+    // counting_iterator
+    test_with_base_iterator<true, oneapi::dpl::counting_iterator<std::int32_t>>();
+
+    // pointer (USM assumed)
+    test_with_base_iterator<true, int*>();
+
+    // create a usm allocated vector
+    sycl::queue q;
+    sycl::usm_allocator<int, sycl::usm::alloc::shared> alloc(q);
+    std::vector<int, sycl::usm_allocator<int, sycl::usm::alloc::shared>> vec(alloc);
+    test_with_base_iterator<TestUtils::__vector_impl_distinguishes_usm_allocator_from_default_v<decltype(vec.begin())>,
+                            decltype(vec.begin())>();
+
+    // custom iter type with legacy is_passed_directly trait defined
+    test_with_base_iterator<true, IDA_iter>();
+
+    // custom iter type with explicit is_passed_directly trait defined as false
+    test_with_base_iterator<false, explicit_non_IDA_iterator>();
+
+    // custom iter type implicitly not device accessible content iterator
+    test_with_base_iterator<false, implicit_non_IDA_iter>();
+
+    // std vector with normal allocator
+    std::vector<int> vec2(10);
+    test_with_base_iterator<false, decltype(vec2.begin())>();
+
+    // test discard_iterator
+    static_assert(oneapi::dpl::is_indirectly_device_accessible_v<oneapi::dpl::discard_iterator> == true,
+                  "is_onedpl_indirectly_device_accessible_iterator is not working correctly for discard iterator");
```

Comment:
We have a public type  - `discard_iterator`,
Probably, should we use  ".....is not working correctly for `discard_iterator`" ?